parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt

```diff
@@ -0,0 +1,158 @@
+package com.parsely.parselyandroid
+
+import androidx.test.core.app.ApplicationProvider
+import java.util.Calendar
+import java.util.TimeZone
+import java.util.Timer
+import org.assertj.core.api.AbstractLongAssert
+import org.assertj.core.api.Assertions.assertThat
+import org.assertj.core.api.Assertions.within
+import org.assertj.core.api.Assertions.withinPercentage
+import org.assertj.core.api.MapAssert
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+
+private typealias Event = MutableMap<String, Any>
+
+@RunWith(RobolectricTestRunner::class)
+@Suppress("UNCHECKED_CAST")
+internal class EngagementManagerTest {
+
+    private lateinit var sut: EngagementManager
+    private val tracker = FakeTracker()
+    private val parentTimer = Timer()
+    private val baseEvent: Event = mutableMapOf(
+        "action" to "heartbeat",
+        "data" to testData
+    )
+
+    @Before
+    fun setUp() {
+        sut = EngagementManager(
+            tracker,
+            parentTimer,
+            DEFAULT_INTERVAL_MILLIS,
+            baseEvent,
+            FakeIntervalCalculator()
+        )
+    }
+
+    @Test
+    fun `when starting manager, then record the correct event after interval millis`() {
+        // when
+        sut.start()
+        sleep(DEFAULT_INTERVAL_MILLIS)
+        val timestamp = now - THREAD_SLEEPING_THRESHOLD
+
+        // then
+        assertThat(tracker.events[0]).isCorrectEvent(
+            withTotalTime = {
+                // Ideally: totalTime should be equal to DEFAULT_INTERVAL_MILLIS
+                isCloseTo(DEFAULT_INTERVAL_MILLIS, withinPercentage(10))
+            },
+            withTimestamp = {
+                // Ideally: timestamp should be equal to System.currentTimeMillis() at the time of recording the event
+                isCloseTo(
+                    timestamp,
+                    within(100L)
+                )
+            }
+        )
+    }
+
+    @Test
+    fun `when starting manager, then schedule task each interval period`() {
```

Comment:
Question (❓): Please help me understand how `start()` works on `EngagementManager` for multiple events. 🙏

FYI: I am asking that because, to my understanding, a new instance of `EngagementManager` is created with just one event, the `baseEvent`, and that is the only event needed to be enqueued. I am struggling a bit to understand the logic here and when we can have multiple of the same events queuing. 🤔

PS: This is related to 51a467463c5f0602100a436f76d21bed2f375db9 .

Reply:
When we call `start` on `EngagementManager`, the `java.utils.TimerTask` is scheduled inside `java.utils.Timer`. The next execution of the `TimerTask` will schedule a subsequent task.

So this test checks if subsequent tasks are executed and if they have correct parameters like e.g. updated total timestamp. Does this answer your questions @ParaskP7 ?

Reply:
Hmmm, yes, I understand how the `TimerTask` will schedule a subsequent task. I just don't understand when multiple events, mostly in production, could be scheduled, can you help me with that? 🙏 

```java
        engagementManager = new EngagementManager(this, timer, DEFAULT_ENGAGEMENT_INTERVAL_MILLIS, event, intervalCalculator);
        engagementManager.start();
```

I see the above happening on production, which on (for example) `startEngagement(...)` creates a new instance of `EngagementManager` with a specific single `event`. How is this manager is able to handle multiple events, when does this happens, this is what I don't understand @wzieba . 🤔

Reply:
>  How is this manager is able to handle multiple events, when does this happens, this is what I don't understand @wzieba . 🤔

When you say "multiple events" do you mean "multiple engagement sessions"? If yes, then the SDK cannot handle it.

Circling back to the original question:

> I am asking that because, to my understanding, a new instance of EngagementManager is created with just one event, the baseEvent, and that is the only event needed to be enqueued.

Yes, the `EngegementManager` is created with just one event, the `baseEvent`. But no - it's not the only event needed to be enqueued. After the initial `baseEvent` is enqueued, enqueuing of the next event is scheduled.

When interval time passes, the SDK, based on `baseEvent`, enqueues a new event with modifications done in `doEnqueue`. It modifies, e.g. `ts`, `inc` or `tt` parameters.

Reply:
> When you say "multiple events" do you mean "multiple engagement sessions"? If yes, then the SDK cannot handle it.

Good to know! 💯 

> Yes, the EngegementManager is created with just one event, the baseEvent. But no - it's not the only event needed to be enqueued. After the initial baseEvent is enqueued, enqueuing of the next event is scheduled.

👍 

> When interval time passes, the SDK, based on baseEvent, enqueues a new event with modifications done in doEnqueue. It modifies, e.g. ts, inc or tt parameters.

That explains it, and apologies for only knowing the basics of the SDK, trying to figure out the rest during reviews... 😊 

So, when `startEngagement(...)` gets triggered, it is always triggered with a `baseEvent`, and then, until `stopEngagement(...)` is called, the same `baseEvent` will be enqueued multiple time, based on the interval, with only some of its parameters being modified. If so, me repeating correctly what you already said, I then think I got the complete picture now, thanks so much for explaining. 🙇

Reply:
> the same baseEvent will be enqueued multiple time, based on the interval, with only some of its parameters being modified

Yes, exactly this 🙌 !